ql/src/java/org/apache/hadoop/hive/ql/plan/ConditionalResolverMergeFiles.java

```diff
@@ -380,32 +432,32 @@ public int getNumFiles() {
   private AverageSize getAverageSize(FileSystem inpFs, Path dirPath) {
     AverageSize error = new AverageSize(-1, -1);
     try {
-      FileStatus[] fStats = inpFs.listStatus(dirPath);
-
-      long totalSz = 0;
-      int numFiles = 0;
-      for (FileStatus fStat : fStats) {
-        Utilities.FILE_OP_LOGGER.debug("Resolver looking at " + fStat.getPath());
-        if (fStat.isDir()) {
-          AverageSize avgSzDir = getAverageSize(inpFs, fStat.getPath());
-          if (avgSzDir.getTotalSize() < 0) {
-            return error;
-          }
-          totalSz += avgSzDir.getTotalSize();
-          numFiles += avgSzDir.getNumFiles();
-        }
-        else {
-          totalSz += fStat.getLen();
-          numFiles++;
-        }
-      }
-
-      return new AverageSize(totalSz, numFiles);
+      return getAverageSize(getFileStatus(inpFs, dirPath));
     } catch (IOException e) {
       return error;
     }
   }
 
+  private List<FileStatus> getFileStatus(FileSystem inputFs, Path path) throws IOException {
+    FileStatus[] fStats = inputFs.listStatus(path);
+    List<FileStatus> result = new ArrayList<>();
+    for (FileStatus stat : fStats) {
+      if (stat.isDirectory()) {
+        result.addAll(getFileStatus(inputFs, stat.getPath()));
+      } else {
+        result.add(stat);
+      }
+    }
+    return result;
+  }
+
+  private AverageSize getAverageSize(List<FileStatus> fileStatusList) {
+    long totalSize = fileStatusList.stream().filter(FileStatus::isFile)
+        .mapToLong(FileStatus::getLen).sum();
+    int numOfFiles = fileStatusList.size();
```

Comment:
that is not accurate as above you have a filter(FileStatus::isFile)

Reply:
Done.